src/libraries/System.Net.Http/src/System/Net/Http/HttpContent.cs

```diff
@@ -254,6 +254,53 @@ internal byte[] ReadBufferedContentAsByteArray()
             return _bufferedContent.ToArray();
         }
 
+        public Stream ReadAsStream() =>
+            ReadAsStream(CancellationToken.None);
+
+        public Stream ReadAsStream(CancellationToken cancellationToken)
+        {
+            CheckDisposed();
+
+            // _contentReadStream will be either null (nothing yet initialized), a Stream (it was previously
+            // initialized in TryReadAsStream/ReadAsStream), or a Task<Stream> (it was previously initialized
+            // in ReadAsStreamAsync).
+
+            if (_contentReadStream == null) // don't yet have a Stream
+            {
+                Stream s = TryGetBuffer(out ArraySegment<byte> buffer) ?
+                    new MemoryStream(buffer.Array!, buffer.Offset, buffer.Count, writable: false) :
+                    CreateContentReadStream(cancellationToken);
+                _contentReadStream = s;
+                return s;
+            }
+            else if (_contentReadStream is Stream stream) // have a Stream
+            {
+                return stream;
+            }
+            else // have a Task<Stream>
+            {
+                // Several options here:
```

Comment:
There are several options how to deal with subsequent calls to `ReadAsStream`, `ReadAsStreamAsync` and result caching in `_contentReadStream`. They're listed in the comments.

In the end, I like 4. the best.

Reply:
If someone calls `ReadAsStreamAsync` and then `ReadAsStream`, there's a usage error. My preference is option 1 (throw on usage error), followed by option 2 (try our best but throw if it would block).

Option 4 would increase the size of the object by 8 bytes, so I would not do that. I don't think the extra memory is worth it to support that weird scenario.

Reply:
Let's do (1).

Reply:
Done